### PR TITLE
HEEDLS-992 Update permissions for pages in my account area

### PR DESCRIPTION
**Pull request by AlexJacksonDS**
### JIRA link
https://softwiretech.atlassian.net/browse/HEEDLS-992

### Description
Fixed the permissions levels for pages in the my account area. Renamed User.GetCentreId to User.GetCentreIdKnownNotNull so that matches the other places where we have potentially null claims.

Did some basic refactoring of the My account index and edit details actions to allow them to load for a centreless user. The content/form posts etc will be fixed in later tickets.

### Screenshots
N/A

-----
### Developer checks
(Leave tasks unticked if they haven't been appropriate for your ticket.)

I have:
- [x] Run the formatter and made sure there are no IDE errors.
- [ ] Written tests for the changes (accessibility tests, unit tests for controller, data services, services, view models, etc)
- [x] Manually tested my work with and without JavaScript. Full manual testing guidelines can be found here: https://softwiretech.atlassian.net/wiki/spaces/HEE/pages/6703648740/Testing
- [ ] Updated/added documentation in Swiki and/or Readme. Links (if any) are below:
  - [doc_1_here](link_1_here)
- [ ] Updated my Jira ticket with information about other parts of the system that were touched as part of the MR and have to be sanity tested to ensure nothing’s broken.
- [x] Scanned over my own MR to ensure everything is as expected.

### CODE REVIEW

**Review by VladNicol — APPROVED** [review]
Looks good to me, thanks for the TODO comments in the code!

File: DigitalLearningSolutions.Web/Helpers/CustomPolicies.cs [hunk 1]
```diff
@@ -18,6 +19,13 @@ public static AuthorizationPolicyBuilder ConfigurePolicyBasicUser(AuthorizationP
             return policy.RequireAssertion(context => context.User.GetUserId() != null);
         }
 
+        public static AuthorizationPolicyBuilder ConfigurePolicyCentreUser(AuthorizationPolicyBuilder policy)
+        {
+            return policy.RequireAssertion(
+                context => context.User.GetUserId() != null && context.User.GetCentreId() != null
```
Comment (DanBloxham-sw) [inline DigitalLearningSolutions.Web/Helpers/CustomPolicies.cs:25]: We've got a number of helper methods defined in the CustomClaimHelper such as `HasCentreAdminPermissions`. Do we want to define similar ones for the IsBasicUser and IsCentreUser scenarios, and apply them across these policies?

File: DigitalLearningSolutions.Web/Controllers/NotificationPreferencesController.cs [hunk 2]
```diff
@@ -15,6 +15,7 @@ namespace DigitalLearningSolutions.Web.Controllers
     [TypeFilter(typeof(ValidateAllowedDlsSubApplication))]
     [SetDlsSubApplication]
     [SetSelectedTab(nameof(NavMenuTab.MyAccount))]
+    [Authorize(Policy = CustomPolicies.CentreUser)]
     public class NotificationPreferencesController : Controller
```
Comment (DanBloxham-sw) [inline DigitalLearningSolutions.Web/Controllers/NotificationPreferencesController.cs:19]: Should we be hiding the button on the My Account page for this?

File: DigitalLearningSolutions.Web/Controllers/MyAccountController.cs [hunk 3]
```diff
@@ -96,18 +97,23 @@ public IActionResult Index(DlsSubApplication dlsSubApplication)
             var centreId = User.GetCentreId();
             var userId = User.GetUserIdKnownNotNull();
             var userEntity = userService.GetUserById(userId)!;
-            var delegateAccount =
-                userEntity.DelegateAccounts.SingleOrDefault(da => da.CentreId == centreId);
+            var delegateAccount = userEntity.GetCentreAccountSet(centreId)?.DelegateAccount;
```
Comment (DanBloxham-sw) [inline DigitalLearningSolutions.Web/Controllers/MyAccountController.cs:100]: I'm getting a warning about a possible NullReferenceException on the userEntity used in this line, despite the forgiving operator being applied above. Same in the Index. Weird.

**Review by DanBloxham-sw — CHANGES_REQUESTED** [review]
Some minor points.

File: DigitalLearningSolutions.Web/Controllers/NotificationPreferencesController.cs (hunk 2 above)
Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Web/Controllers/NotificationPreferencesController.cs:19]: Content changes are coming in the later tickets that I left TODOs for.

File: DigitalLearningSolutions.Web/Helpers/CustomPolicies.cs (hunk 1 above)
Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Web/Helpers/CustomPolicies.cs:25]: I'm not sure that it adds much? I don't think these levels will ever be used outside policies, unlike the HasCentreAdminPermissions/HasCentreManagerPermissions extension methods.

Comment (DanBloxham-sw) [reply DigitalLearningSolutions.Web/Helpers/CustomPolicies.cs:25]: Fair enough.

File: DigitalLearningSolutions.Web/Controllers/MyAccountController.cs (hunk 3 above)
Comment (DanBloxham-sw) [reply DigitalLearningSolutions.Web/Controllers/MyAccountController.cs]: Yeah that looks to have solved it.

**Review by DanBloxham-sw — APPROVED** [review]
Looks good.

**Pull request merged**